walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,87 @@
 
 package com.google.sps;
 
+import java.sql.Time;
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  private static final Collection<Event> NO_EVENTS = Collections.emptySet();
+  private static final Collection<String> NO_ATTENDEES = Collections.emptySet();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> ranges = new ArrayList<>();
+    if((events == NO_EVENTS&& request.getDuration()<=TimeRange.WHOLE_DAY.duration()) 
+      || request.getAttendees() == NO_ATTENDEES){
+      ranges.add(TimeRange.WHOLE_DAY);
+      return ranges;
+    }else if(request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return ranges;
+    }
+    Collection<String> attendees = request.getAttendees();
+    for(Event event: events){
+      Collection<String> eAttendees = event.getAttendees();
+      for(String attendee: attendees){
+        if(eAttendees.contains(attendee)||eAttendees.isEmpty()){
+          ranges.add(event.getWhen());
+        }
+      }
+    }
+    if(ranges.size()==0){
+      ranges.add(TimeRange.WHOLE_DAY);
+      return ranges;
+    }
+    Collection<TimeRange> overlapedRanges = new ArrayList<>();
+    if(ranges.size()>1){
+      ranges.sort(TimeRange.ORDER_BY_START);
+      TimeRange r1 = ranges.get(0);
+      for(int i = 1; i < ranges.size();i++){
+        TimeRange r2 = ranges.get(i);
+        if(r1.overlaps(r2)){
+          if(r1.contains(r2)){
+            overlapedRanges.add(TimeRange.fromStartEnd(r1.start(), r1.end(), false));
+          }else{
+            overlapedRanges.add(TimeRange.fromStartEnd(r1.start(), r2.end(), false));
+          }
+        }else{
+          overlapedRanges.add(TimeRange.fromStartEnd(r1.start(), r1.end(), false));
+          if(i+1 >= ranges.size()){
+            overlapedRanges.add(TimeRange.fromStartEnd(r2.start(), r2.end(), false));
+          }else{
+            r1 = r2;
+          }
+        }
+      }
+    }else if(ranges.size()==1){
+      overlapedRanges.add(TimeRange.fromStartEnd(ranges.get(0).start(), ranges.get(0).end(), false));
+    }
+    Collection<TimeRange> validRanges = getValidTimes(overlapedRanges,request.getDuration());
+  return validRanges;
+  }
+
+  public Collection<TimeRange> getValidTimes(Collection<TimeRange> ranges, long duration){
```

Comment:
What are your assumptions on this method? That ranges are sorted and disjoint (the elements have no overlap)? Write your assumptions in a comment.